### PR TITLE
Fix share permission checks

### LINKED ISSUE #13193: can not upload changed shared file via syncclient with only change permissions
### Steps to reproduce
1. log in with admin account
2. create a user "user1"
3. create a folder "sharetest"
4. share the folder "sharetest" with user1, but with only "change" permission
5. go into "sharetest" and create a text file "permissiontest.txt"
6. set syncclient account for user1
7. open the local "data" folder for user1
8. go into "sharetest" and open "permissiontest.txt"
9. make some changes of this text file and save it
10. wait for sync of syncclient (ERROR1)
11. rename the text file to "sharetest.txt"
12. wait for sync of syncclient (ERROR2)

### Expected behaviour
The changed file can be uploaded, and "admin" can see the changes of the file.

### Actual behaviour
ERROR1:
The changed file can not be uploaded due to an error:
Error downloading http://192.168.1.80/owncloud704/remote.php/webdav/sharetest/permissiontest.txt - server replied: Internal Server Error (Could not write file contents) 

ERROR2: 
sharetest/sharetest.txt--Not allowed because you don't have permission to add files in that directory          
sharetest/permissiontest.txt--Not allowed to remove, restoring          
sharetest/sharetest.txt--Not allowed because you don't have permission to add files in that directory    
      
User1 can find 2 text files in local sync client data/sharetest folder, permissiontest.txt (the origin one) and sharetest.txt (the changed one), but neither user1 or admin can get all these changes in the web ownCloud

### Server configuration
**Operating system:** Ubuntu 13.10

**Web server:** apache 2.4.6 (Ubuntu)

**Database:** SQLite

**PHP version:** 5.5.3-1ubuntu2.3

**ownCloud version:** 7.0.4 & master

**Updated from an older ownCloud or fresh install:** owncloud 7.0.3

**List of activated apps:** standard apps of owncloud

**The content of config/config.php:**

```
Insert your config.php content here
(Without the database password and passwordsalt)
$CONFIG = array (
  'instanceid' => 'oc6d9cda6e26',

  'datadirectory' => '/var/ocdata/.../owncloud/data',
  'dbtype' => 'sqlite3',
  'version' => '7.0.4.2',
  'installed' => true,
  'theme' => '',
  'maintenance' => false,
  'maxZipInputSize' => 838860800,
  'allowZipDownload' => true,
  'loglevel' => '0',
  'trusted_domains' => 
  array (
    0 => '192.168.1.80',
  ),

);
```

**Are you using external storage, if yes which one:** no
**Are you using encryption:** yes

### Client configuration
**Browser:** firefox 34.0.5

**Operating system:** ubuntu 13.10

#### Web server error log
```

```

#### ownCloud log (data/owncloud.log)
```

```

#### Browser log
```

```

### PR DESCRIPTION
**Pull request by PVince81**
Fixes https://github.com/owncloud/core/issues/13193

@icewind1991 @schiesbn 
- [x] add unit test

### CODE REVIEW

**Comment by icewind1991** [comment]
It should still check for updatable permissions

**Comment by PVince81** [comment]
I added more unit tests for permissions, it confirmed what you meant.
Also, it fixes additional permission issue.

I haven't tested this extensively yet, need to do that later.
Setting to 8.0.1

**Comment by DeepDiver1975** [comment]
pr is against master -> milestone 8.1

**Comment by PVince81** [comment]
@owncloud-bot retest this please

**Comment by PVince81** [comment]
Trashbin test side-effect, nice... Running those tests alone work.

- [ ] fix unit tests

**Comment by PVince81** [comment]
Commenting out the rename test in the test removes the weird side effect: https://github.com/owncloud/core/pull/13802/files#diff-c4b4f36fc675d709a1256508ee31c76fR221

Needs further research...

**Comment by PVince81** [comment]
Okay, go it. The "pre_rename" hook disables the trash and "post_rename" reenables it.
But here since rename failed, the "post_rename" is never called... and I'm not sure we want to always call "post_rename" on error. Need to find a better way.

**Comment by PVince81** [comment]
@owncloud-bot retest this please

**Comment by PVince81** [comment]
Ah yes, these were the tests:
```
There were 3 failures:

1) OCA\Files_trashbin\Tests\Storage\Storage::testSingleStorageDelete
Failed asserting that 0 matches expected 1.

/srv/www/htdocs/owncloud/apps/files_trashbin/tests/storage.php:80

2) OCA\Files_trashbin\Tests\Storage\Storage::testCrossStorageDelete
Failed asserting that 0 matches expected 1.

/srv/www/htdocs/owncloud/apps/files_trashbin/tests/storage.php:106

3) OCA\Files_trashbin\Tests\Storage\Storage::testDeleteVersions
Failed asserting that 0 matches expected 1.

/srv/www/htdocs/owncloud/apps/files_trashbin/tests/storage.php:131
```

**Comment by PVince81** [comment]
#### Test plan (to be tested with WebDAV and Web UI)
- [ ] create only
- [ ] change only
- [ ] delete only
- [ ] create+change
- [ ] change+delete
- [ ] create+change+delete

**Comment by PVince81** [comment]
(yes, some combinations are a bit ridiculous, please vote for merging of permissions here https://github.com/owncloud/core/issues/12970

**Comment by PVince81** [comment]
@icewind1991 @DeepDiver1975 @schiesbn @nickvergessen please review

**Comment by PVince81** [comment]
and maybe @MorrisJobke :smile:

**Comment by PVince81** [comment]
Rebased onto master (there was an issue related to the obsoleted oc_locks)

**Comment by PVince81** [comment]
Looks like we're not done yet... more inconsistencies:
- [ ] BUG: move file out of "create+delete": web UI works but WebDAV says forbidden
- [ ] BUG: move file out of "delete": web UI works but WebDAV says forbidden

**Comment by PVince81** [comment]
- [ ] BUG (?): move a file "a.txt" into folder "update" (with only update permissions), and "update/a.txt" already exists. Moving a file there is technically overwriting, which is an update, not create. This doesn't work (WebDAV)

**Comment by PVince81** [comment]
Wrote a smashbox test suite to make it easier to test the WebDAV part: https://github.com/owncloud/smashbox/pull/38

**Comment by PVince81** [comment]
This permission thing and interpretations is a real nightmare.

I just found out again that if you try to overwrite a file by moving, Sabre will automatically delete the target (which then requires DELETE permission). But the reason we have the UPDATE permission is to allow people to overwrite files, so that permission should be enough in itself.

It is not possible to add any clean workarounds for this, so I might end up agreeing that "overwrite" needs UPDATE + DELETE permissions... for both upload and rename.

**Comment by PVince81** [comment]
- [ ] BUG: web UI shows rename icon for (update+delete and update) but rename only works if create+update is granted

**Comment by PVince81** [comment]
Damn, just compared with master and it looks like so far we always allowed "rename in place" if the update permission was granted.
I'll try to reallow that without breaking the other permissions.

**Comment by PVince81** [comment]
- [ ] fix sabre unit tests for the updated move permission check

File: apps/files_sharing/tests/sharedstorage.php [hunk 1]
```diff
@@ -199,6 +199,154 @@ function testGetPermissions() {
...
+		$this->assertFalse($user2View->unlink($this->folder . '/existing.txt'));
+
+		//cleanup
+		self::loginHelper(self::TEST_FILES_SHARING_API_USER1);
+		$result = \OCP\Share::unshare('folder', $fileinfoFolder['fileid'], \OCP\Share::SHARE_TYPE_USER,
+			self::TEST_FILES_SHARING_API_USER2);
+		$this->assertTrue($result);
+	}
+
+	function testFopenWithDeleteOnlyPermission() {
```
Comment (nickvergessen) [inline apps/files_sharing/tests/sharedstorage.php]: please add `public` next time ;)

**Comment by PVince81** [comment]
Fixed the unit tests. This is ready for review.

I'll look into the "overwrite read-only shared file" separately.

**Comment by PVince81** [comment]
@icewind1991 do you know why the permission check logic is duplicated ? One is here https://github.com/owncloud/core/blob/master/lib/private/connector/sabre/objecttree.php#L177 and the other part is in the shared storage.
Maybe we should rename the part in sabre ? (at least I got smashbox tests now to confirm it still works)

**Comment by icewind1991** [comment]
Looks like the sabre check can go

**Comment by icewind1991** [comment]
looks good :+1:

**Comment by PVince81** [comment]
I'll remove it separately. I tried it locally but ran into complications (aka more work required)

**Comment by MorrisJobke** [comment]
Tested and works :+1:

**Comment by PVince81** [comment]
@karlitschek backport ?

**Comment by karlitschek** [comment]
please backport

**Comment by PVince81** [comment]
stable8: https://github.com/owncloud/core/pull/15246

**Comment by karlitschek** [comment]
thx

**Pull request merged**